### ISSUE TITLE
ConfigureAwait does nothing?

### ISSUE THREAD

**Issue by gkhanna79** [dotnet/corefx#16973]
@borrrden commented on [Thu Mar 09 2017](https://github.com/dotnet/coreclr/issues/10082)

I ran into this unusual problem when running a unit test with xUnit.  I noticed the thread before I `await`ed an `async` call and the thread when I was done were different.  Ok, I guess this makes sense after reading about `SynchronizationContext`.  However, the weird part is that things are continuing on the same thread that the `async` task was running on.  That thread is reserved for processing async jobs, and so any blocking after the await in the original test code will cause that processing to lock.

My implementation of this is [here](https://github.com/couchbase/couchbase-lite-net/blob/feature/api-v2/src/Couchbase.Lite/Support/SerialQueue.cs#L136).  Roughly, it adds some logic to a queue and sets the result of a `TaskCompletionSource` when it is done.  In terms of what I am saying above, the logic gets executed in `ProcessAsync` and then the test code continues **on the same thread that ProcessAsync is running**

I inspected the `SynchronizationContext` of the thread at start, and determined that it is an XUnit library provided instance of [AsyncTaskMethodContext](https://github.com/xunit/xunit/blob/master/src/xunit.execution/Sdk/AsyncTestSyncContext.cs).

If I understand correctly, without `ConfigureAwait(false)` the async state machine should post back to the original context when it is finished with its logic.  In that case, this instance (I inspected `innerContext` to be null) will call [XunitWorkerThread.QueueUserWorkItem](https://github.com/xunit/xunit/blob/vs2017/src/common/XunitWorkerThread.cs#L29) which just adds to a default constructure `TaskFactory` instance.  That seems to me that it should run on yet another thread and not the same one as before since the one from before is still executing stuff.

I thought maybe the default changed to `ConfigureAwait(false)` so I tried running `ConfigureAwait(true)` but the behavior was the same.  Is this a bug or the new behavior?

---

@stephentoub commented on [Fri Mar 10 2017](https://github.com/dotnet/coreclr/issues/10082#issuecomment-285661409)

When you write code like:
```C#
var result = await t;
```
that logically translates into something like:
```C#
var $awaiter = t.GetAwaiter();
if (!$awaiter.IsCompleted)
{
    ... // stuff to hook up a continuation
    return;
}
Label:
var result = $awaiter.GetResult();
```
so if the thing you await is already completed by the time you await it, you will continue running synchronously on the same thread.  This has always been the behavior, and ConfigureAwait doesn't influence that, i.e. ConfigureAwait doesn't force the await to always complete asynchronously, but rather controls the continuation behavior if it does complete asynchronously.

---

@stephentoub commented on [Fri Mar 10 2017](https://github.com/dotnet/coreclr/issues/10082#issuecomment-285662040)

Separately, in your code, Task.Factory.StartNew by default targets the current TaskScheduler, and it's up to the current TaskScheduler how to interpret and handle TaskCreationOptions.LongRunning.  If your intent here is to force a new thread to be created as happens for LongRunning with the default scheduler, you should explicitly pass TaskScheduler.Default as the scheduler argument to StartNew; otherwise, if the context in which you're running has a TaskScheduler set up, you may not get the behavior you want (of course, maybe that's exactly what you want, but just wanted to call it out).

**Comment by stephentoub** [comment]
@borrrden, I'm going to close this, as I don't believe there's a bug in ConfigureAwait here.  If you're still seeing problems that you believe are due to the implementation, we can re-open this.  Thanks.

**Comment by borrrden** [comment]
Ok, but just to make sure I said things clearly and I understand the response, the following is expected behavior?

1) Start a unit test on thread 8 (for example)
2) await a call to the async processor (queue a `Func<T>` for execution and return a `Task` representing its status)
3) async processor (ProcessAsync method) runs `Func<T>` on thread 12 (for example) and appropriately sets the `Task` result
4) Unit test continues on thread 12 (I would have expected either thread 8, or a completely different thread given the behavior of the installed synchronization context)
5) Unit test blocks waiting for a reset event (for example) and ProcessAsync is now locked.

[Here](https://github.com/couchbase/couchbase-lite-net/blob/feature/api-v2/src/Couchbase.Lite/Support/SerialQueue.cs#L101-L106) is my hacky workaround to stop this from happening but I don't like it...

Thanks for the reminder about `TaskScheduler.Default` too!

**Comment by borrrden** [comment]
I've come up with a better way to approach this I think.  Instead of the hacky workaround, I should capture the current sync context and post the `tcs.SetResult` to that instead (or use a new one if `Current` is `null` which will default to the hacky workaround).  I think I understand what is going on now anyway.  Thanks for your time!

**Comment by stephentoub** [comment]
> the following is expected behavior?

It depends on a whole bunch of things.  It would depend, for example, on what SynchronizationContext is present as Current on thread 8, what TaskScheduler.Current is on thread 8, etc.  You could keep executing on thread 12 even if there was a synchronization context, if the context wasn't affinitized to a specific thread, for example, and just enforced certain rules about execution, such as only allowing one associated work item to be running at a time.

**Comment by borrrden** [comment]
In the above case SynchronizationContext on thread 8 was what I mentioned earlier, and TaskScheduler.Current on thread 8 was TaskScheduler.Default.

So I will rephrase:  The behavior I observed is possible, and not considered a problem?

I'm satisfied with the way it is working now that I replaced the hack above with posting back to the synchronizationcontext from the thread that called DispatchAsync.  This issue can remain closed.
